### ISSUE TITLE
[1.11.1] Docker daemon hanging on startup

### ISSUE THREAD

**Issue by leedm777** [docker/docker#23131]
**Output of `docker version`:**

```
Client:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:30:23 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:30:23 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 2
 Running: 2
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.11.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 30
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 3.13.0-86-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.954 GiB
Name: ip-172-18-0-56
ID: 53GT:ASJT:ASAC:OYKP:7T7S:RF6Q:H4L2:VZGK:QB2N:W56A:X6WC:3P6I
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): true
 File Descriptors: 18
 Goroutines: 41
 System Time: 2016-05-31T16:38:06.468909009Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

AWS; docker is hanging on initial launch from AMI. AMI is built from Ubuntu Trusty, with linux-image-extras and docker installed.

**Steps to reproduce the issue:**
1. Launch an AMI with Docker 1.11.1 installed

**Describe the results you received:**

Docker is hanging on startup. Any docker commands that communicate with the daemon (`ps`, `info`, `run`, `pull`, ...) hang indefinitely.

```
time="2016-05-31T16:29:27.742468971Z" level=debug msg="docker group found. gid: 999" 
time="2016-05-31T16:29:27.742553101Z" level=debug msg="Listener created for HTTP on unix (/var/run/docker.sock)
" 
/var/run/docker.sock is up
time="2016-05-31T16:29:28.440531154Z" level=info msg="New containerd process, pid: 1619\n" 
time="2016-05-31T16:29:28.759224749Z" level=debug msg="containerd connection state change: CONNECTING" 
time="2016-05-31T16:29:28.759533205Z" level=debug msg="containerd connection state change: TRANSIENT_FAILURE" 
time="2016-05-31T16:29:31.759889842Z" level=debug msg="containerd connection state change: TRANSIENT_FAILURE" 
DEBU[0000] containerd: read past events                  count=0
DEBU[0000] containerd: supervisor running                cpus=1 memory=2000 runtime=docker-runc runtimeArgs=[] 
stateDir=/run/containerd
DEBU[0000] containerd: grpc api on /var/run/docker/libcontainerd/docker-containerd.sock 
time="2016-05-31T16:29:34.760226204Z" level=debug msg="containerd connection state change: TRANSIENT_FAILURE" 
time="2016-05-31T16:29:34.760298247Z" level=info msg="previous instance of containerd still alive (1619)" 
```

**Describe the results you expected:**

Docker should start normally.

```
/var/run/docker.sock is up
time="2016-05-31T16:36:13.336502737Z" level=debug msg="docker group found. gid: 999" 
time="2016-05-31T16:36:13.336649784Z" level=debug msg="Listener created for HTTP on unix (/var/run/docker.sock)
" 
time="2016-05-31T16:36:13.340959840Z" level=info msg="New containerd process, pid: 2098\n" 
DEBU[0000] containerd: read past events                  count=0
DEBU[0000] containerd: supervisor running                cpus=1 memory=2000 runtime=docker-runc runtimeArgs=[] 
stateDir=/run/containerd
DEBU[0000] containerd: grpc api on /var/run/docker/libcontainerd/docker-containerd.sock 
time="2016-05-31T16:36:13.348085025Z" level=debug msg="containerd connection state change: READY" 
time="2016-05-31T16:36:13.348764443Z" level=debug msg="Using default logging driver json-file" 
time="2016-05-31T16:36:13.348807782Z" level=debug msg="Golang's threads limit set to 28350" 
time="2016-05-31T16:36:13.397853157Z" level=debug msg="received past containerd event: &types.Event{Type:\"live
\", Id:\"\", Status:0x0, Pid:\"\", Timestamp:0x574dbd7d}" 
time="2016-05-31T16:36:14.245629031Z" level=info msg="[graphdriver] using prior storage driver \"aufs\"" 
time="2016-05-31T16:36:14.245735836Z" level=debug msg="Using graph driver aufs" 
time="2016-05-31T16:36:17.775448474Z" level=info msg="Graph migration to content-addressability took 0.00 secon
ds" 
time="2016-05-31T16:36:17.775596758Z" level=debug msg="Option DefaultDriver: bridge" 
time="2016-05-31T16:36:17.775648036Z" level=debug msg="Option DefaultNetwork: bridge" 
time="2016-05-31T16:36:18.177452410Z" level=warning msg="Running modprobe bridge br_netfilter failed with messa
ge: modprobe: WARNING: Module br_netfilter not found.\ninsmod /lib/modules/3.13.0-86-generic/kernel/net/llc/llc.ko \ninsmod /lib/modules/3.13.0-86-generic/kernel/net/802/stp.ko \ninsmod /lib/modules/3.13.0-86-generic/kernel/net/bridge/bridge.ko \n, error: exit status 1" 
time="2016-05-31T16:36:19.002029744Z" level=info msg="Firewalld running: false" 
time="2016-05-31T16:36:19.176005296Z" level=debug msg="/sbin/iptables, [--wait --version]" 
<snip/>
```

**Additional information you deem important (e.g. issue happens only occasionally):**

* The hang happens frequently, but not every time we launch an AMI.
* After the instance is launched, `restart docker` consistently gets the daemon running.
* I tried sending `SIGUSR1` to the daemon to get a stack trace, but nothing showed in the logs. I confirmed that the command I used (`killall -USR1 docker`) did generate a stack dump once Docker was properly running.

**Comment by mlaventure** [comment]
From the log, here what seems to happen:
 - `containerd` gets spawned by `docker`, but it takes too long to get its `grpc` socket ready.
 - after 3 `TRANSIENT_FAILURE` (1 every 3 seconds), `docker` decides to kill the current `containerd` and restart it.
 - it looks that by the time we call the routine to start `containerd` the old one is somehow still alive (how did it survive `SIGKILL`?)
 - Thus, we return without created a new `containerd` instance and I guess after that the original one finally dies.

I don't know where docker is stuck afterwards though. But given that it doesn't dump a stack trace, it must be before we call `setupDumpStackTrap()` and from what I can see, there's nothing before that should be blocking.

For the case of `containerd` being alive after we tried to kill it, I'll open a PR to wait for the process to disappear before trying to re-spawn it.

**Comment by danbeaulieu** [comment]
FWIW we are seeing this as well on the same exact setup, except we are using docker 1.11.0.

**Comment by danbeaulieu** [comment]
As an update I think I have reproduction steps (though not the root cause):

1) Create an instance from an Ubuntu cloud image (ami-0f8bce65) using any instance family with ephemeral drives that also aren't i3 or r3 family types
2) Install docker
3) verify xvdb is mounted to /mnt in /etc/fstab 
4) create an AMI from this instance
5) launch an r3 instance with your new ami

Mounting the ephemeral drive fails because r3 instances don't have a filesystem on the ephemeral SSDs. This is done via cloud-init. For whatever reason this leaves the system in a state where docker operations hang.

**Comment by danbeaulieu** [comment]
This happens pretty consistently. I've added an strace while a docker ps was hung for several minutes. The strace itself is for about a minute, not sure if there is anything useful in there.

@mlaventure does it make sense to open this back up? I'm not sure how waiting for containerd is the fix for this when it looks like another lock contention issue.

[stuck-docker-strace.txt](https://github.com/docker/docker/files/305128/stuck-docker-strace.txt)

**Comment by mlaventure** [comment]
@danbeaulieu can you also post the daemon log that goes with this strace?

**Comment by danbeaulieu** [comment]
@mlaventure it looks exactly like the docker log @leedm777 posted in the original issue. When I see it again I'll update if you'd like.

**Comment by hridyeshpant** [comment]
We are also facing this issue with Docker version 1.11.2, build b9f10c9/1.11.2. 
Also this is not happening every time.

Docker log :
time="2016-08-19T04:11:53Z" level=debug msg="containerd: supervisor running" cpus=2 memory=3954 runtime=docker-runc runtimeArgs=[] stateDir="/run/containerd"
time="2016-08-19T04:11:53Z" level=debug msg="containerd: grpc api on /var/run/docker/libcontainerd/docker-containerd.sock"
time="2016-08-19T04:11:54.217522350Z" level=debug msg="containerd connection state change: TRANSIENT_FAILURE"  
time="2016-08-19T04:11:54.217605525Z" level=info msg="previous instance of containerd still alive (2782)”

During this time i try to collect strace of docker process , and i see below things

**Strace of docker process:** 
clock_gettime(CLOCK_MONOTONIC, {1270, 323459387}) = 0
futex(0x2195420, FUTEX_WAKE, 1)         = 1
futex(0x2195360, FUTEX_WAKE, 1)         = 1
clock_gettime(CLOCK_MONOTONIC, {1270, 323682382}) = 0
futex(0xc820026e90, FUTEX_WAKE, 1)      = 1
clock_gettime(CLOCK_REALTIME, {1471581041, 742648765}) = 0
clock_gettime(CLOCK_REALTIME, {1471581041, 742694606}) = 0
clock_gettime(CLOCK_REALTIME, {1471581041, 742744884}) = 0
clock_gettime(CLOCK_REALTIME, {1471581041, 742786631}) = 0
clock_gettime(CLOCK_REALTIME, {1471581041, 742836422}) = 0
clock_gettime(CLOCK_REALTIME, {1471581041, 742881237}) = 0
socket(PF_LOCAL, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0) = 5
setsockopt(5, SOL_SOCKET, SO_BROADCAST, [1], 4) = 0
connect(5, {sa_family=AF_LOCAL, sun_path="/var/run/docker/libcontainerd/docker-containerd.sock"}, 55) = -1 ECONNREFUSED (Connection refused)
close(5)                                = 0
clock_gettime(CLOCK_REALTIME, {1471581041, 743283432}) = 0
futex(0xc820026e90, FUTEX_WAKE, 1)      = 1
clock_gettime(CLOCK_REALTIME, {1471581041, 743401939}) = 0
clock_gettime(CLOCK_MONOTONIC, {1270, 324617172}) = 0
clock_gettime(CLOCK_MONOTONIC, {1270, 324657679}) = 0
futex(0xc820026a10, FUTEX_WAKE, 1)      = 1
clock_gettime(CLOCK_MONOTONIC, {1270, 324829553}) = 0
clock_gettime(CLOCK_MONOTONIC, {1270, 324877163}) = 0
futex(0x2194dd8, FUTEX_WAIT, 0, {2, 999930531}) = -1 ETIMEDOUT (Connection timed out)
clock_gettime(CLOCK_MONOTONIC, {1273, 325106116}) = 0
futex(0x2195420, FUTEX_WAKE, 1)         = 1
futex(0x2195360, FUTEX_WAKE, 1)         = 1
clock_gettime(CLOCK_MONOTONIC, {1273, 325287796}) = 0
futex(0xc820026e90, FUTEX_WAKE, 1)      = 1
futex(0xc820026a10, FUTEX_WAKE, 1)      = 1
clock_gettime(CLOCK_MONOTONIC, {1273, 325585451}) = 0
clock_gettime(CLOCK_MONOTONIC, {1273, 325632618}) = 0
futex(0x2194dd8, FUTEX_WAIT, 0, {2, 999889938}) = -1 ETIMEDOUT (Connection timed out)
clock_gettime(CLOCK_MONOTONIC, {1276, 325813369}) = 0

**Comment by mlaventure** [comment]
Just checked `1.11.2` code, it seems the fix (https://github.com/docker/docker/pull/23148) only went into `1.12.0`

**Comment by robparrott** [comment]
We are seeing this as well and would love to see the fix back ported to at least 1.11.x
